File: clients/android-app/app/src/main/java/com/lijialab/dictionary/DictionaryModels.kt
package com.lijialab.dictionary

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class DictionaryRequest(
    val word: String,
    val section: String
)

@Serializable
data class DictionaryEntry(
    val success: Boolean? = null,
    val error: String? = null,
    val headword: String,
    val entries: List<DictionaryEntryVariant> = emptyList()
) {
    val primaryEntry: DictionaryEntryVariant?
        get() = entries.firstOrNull()

    val senses: List<DictionarySense>
        get() = primaryEntry?.meaningsSummary.orEmpty().flatMap { meaning ->
            meaning.senses.map { sense -> sense.copy(partOfSpeech = meaning.partOfSpeech) }
        }
}

@Serializable
data class DictionaryEntryVariant(
    val pronunciation: String? = null,
    @SerialName("meanings_summary")
    val meaningsSummary: List<DictionaryMeaning> = emptyList()
)

@Serializable
data class DictionaryMeaning(
    @SerialName("part_of_speech")
    val partOfSpeech: String? = null,
    val senses: List<DictionarySense> = emptyList()
)

@Serializable
data class DictionarySense(
    val definition: String? = null,
    val example: String? = null,
    @SerialName("part_of_speech")
    val partOfSpeech: String? = null
)

File: clients/windows-native/src/AdvancedDictionaryWindows/DictionaryModels.cs
using System.Text.Json.Serialization;

namespace AdvancedDictionaryWindows;

public sealed record DictionaryRequest(string Word, string Section);

public sealed class DictionaryEntry
{
    public bool? Success { get; set; }
    public string? Error { get; set; }
    public string Headword { get; set; } = "";
    public List<DictionaryEntryVariant> Entries { get; set; } = new();

    public DictionaryEntryVariant? PrimaryEntry => Entries.FirstOrDefault();

    public IEnumerable<DictionarySense> Senses => PrimaryEntry?.MeaningsSummary.SelectMany(meaning =>
        meaning.Senses.Select(sense => sense with { PartOfSpeech = meaning.PartOfSpeech })) ?? Enumerable.Empty<DictionarySense>();
}

public sealed class DictionaryEntryVariant
{
    public string? Pronunciation { get; set; }

    [JsonPropertyName("meanings_summary")]
    public List<DictionaryMeaning> MeaningsSummary { get; set; } = new();
}

public sealed class DictionaryMeaning
{
    [JsonPropertyName("part_of_speech")]
    public string? PartOfSpeech { get; set; }

    public List<DictionarySense> Senses { get; set; } = new();
}

public sealed record DictionarySense(
    string? Definition,
    string? Example,
    [property: JsonPropertyName("part_of_speech")] string? PartOfSpeech
);
